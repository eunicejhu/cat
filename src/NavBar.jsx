import React from "react";
import { Link } from "react-router-dom";

const NavBar = () => (
  <nav>
    <li>
      <Link to="/">Home</Link>
    </li>
  </nav>
);

export default NavBar;
