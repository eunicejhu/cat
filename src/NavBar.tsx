import React from "react";
import { Link } from "react-router-dom";
import styled from "styled-components";
import SignOutButton from "./components/SignOut";

import { useTheme, Themes, Mode } from "./components/Theme";

import {
    LANDING,
    NOTIFICATIONS,
    SIGN_IN,
    HOME,
    ACCOUNT,
    ADMIN,
} from "./constants/routes";

interface StyledLinkProps {
    themes: Themes;
    mode: Mode;
}
const StyledLink = styled(Link)<StyledLinkProps>`
    ${({ themes, mode }) => {
        return ` --btn-text-hover:  ${themes[mode].btnTextHover};
            --btn-bg-hover: ${themes[mode].btnBackgroundHover};
            --color-bg: ${themes[mode].background} ;
            --color-text:${themes[mode].text};
            --spacing:${themes[mode].spacing};
            `;
    }}
    && {
        text-decoration: none;
        color: var(--color-text);
        background-color: var(--color-bg);
        border: none;
        padding: var(--spacing);
    }

    &&:hover {
        color: var(--btn-text-hover);
        cursor: pointer;
        background-color: var(--btn-bg-hover);
    }
`;

const LinkWithTheme = ({ to, text }: { to: string; text: string }) => {
    const { themes, mode } = useTheme();
    return (
        <li>
            <StyledLink themes={themes} mode={mode} to={to}>
                {text}
            </StyledLink>
        </li>
    );
};

const NavBar = () => {
    return (
        <nav>
            <LinkWithTheme to={LANDING} text="Landing" />
            <LinkWithTheme to={SIGN_IN} text="Sign In" />
            <LinkWithTheme to={HOME} text="Home" />
            <LinkWithTheme to={ACCOUNT} text="Account" />
            <LinkWithTheme to={ADMIN} text="Admin" />
            <LinkWithTheme to={NOTIFICATIONS} text="Notifications" />
            <li>
                <SignOutButton />
            </li>
        </nav>
    );
};

export default NavBar;
