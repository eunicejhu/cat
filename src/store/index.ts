import { configureStore } from "@reduxjs/toolkit";
import postsReducer, { Post } from "../features/posts/postsSlice";
import usersReducer from "../features/users/usersSlice";
import notificationsReducer, {
  Notification,
} from "../features/notifications/notificationsSlice";
import { useDispatch } from "react-redux";

const store = configureStore({
  reducer: {
    posts: postsReducer,
    users: usersReducer,
    notifications: notificationsReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();

export type State = ReturnType<typeof store.getState>;
// selectors
export const selectAllPosts = (state: State): Post[] => state.posts.data;
export const selectPostById = (
  state: State,
  postId: string
): Post | undefined => state.posts.data.find((post) => post.id === postId);
export const selectAllNotifications = (state: State): Notification[] =>
  state.notifications;

export default store;
