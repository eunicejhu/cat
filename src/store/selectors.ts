// selectors
import { State } from "./index";
import { Notification } from "../features/notifications/notificationsSlice";
import { Post } from "../features/posts/postsSlice";
export const selectAllPosts = (state: State): Post[] => state.posts.data;
export const selectPostById = (
  state: State,
  postId: string
): Post | undefined => state.posts.data.find((post) => post.id === postId);
export const selectAllNotifications = (state: State): Notification[] =>
  state.notifications;
