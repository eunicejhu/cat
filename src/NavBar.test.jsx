import React from "react";
import { render, fireEvent } from "@testing-library/react";
import NavBar from "./NavBar";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import store from "./store/index";

const withBrowserRouterAndStore = (ui, route = "/initial") => {
    window.history.pushState({}, "", route);
    return (
        <BrowserRouter>
            <Provider store={store}>{ui}</Provider>
        </BrowserRouter>
    );
};

describe("NavBarNonAuth", () => {
    it("render correctly", () => {
        const ui = withBrowserRouterAndStore(<NavBar authUser={null} />);
        const { getByText } = render(ui);
        expect(getByText(/Landing/g)).toBeInTheDocument();
        expect(getByText(/Sign In/g)).toBeInTheDocument();
    });
});

describe("NavBarAuth", () => {
    it("renders correctly", () => {
        const ui = withBrowserRouterAndStore(<NavBar authUser={{}} />);
        const { asFragment } = render(ui);
        expect(asFragment()).toMatchSnapshot();
    });

    it("click Home, location pathname set to /home", () => {
        const ui = withBrowserRouterAndStore(<NavBar authUser={{}} />);
        const { getByText } = render(ui);
        fireEvent.click(getByText(/Home/i));
        expect(window.location.pathname).toBe("/home");
    });

    it("click Notifications, location pathname set to /notifications", () => {
        const ui = withBrowserRouterAndStore(<NavBar authUser={{}} />);
        const { getByText } = render(ui);
        fireEvent.click(getByText(/Notifications/i));
        expect(window.location.pathname).toBe("/notifications");
    });
});
