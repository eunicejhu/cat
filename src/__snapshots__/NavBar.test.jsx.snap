// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBarAuth renders correctly 1`] = `
<DocumentFragment>
  <nav>
    <li>
      <a
        class="sc-gsTCUz flwUht"
        href="/"
        mode="pink"
        themes="[object Object]"
      >
        Landing
      </a>
    </li>
    <li>
      <a
        class="sc-gsTCUz flwUht"
        href="/home"
        mode="pink"
        themes="[object Object]"
      >
        Home
      </a>
    </li>
    <li>
      <a
        class="sc-gsTCUz flwUht"
        href="/account"
        mode="pink"
        themes="[object Object]"
      >
        Account
      </a>
    </li>
    <li>
      <a
        class="sc-gsTCUz flwUht"
        href="/admin"
        mode="pink"
        themes="[object Object]"
      >
        Admin
      </a>
    </li>
    <li>
      <a
        class="sc-gsTCUz flwUht"
        href="/notifications"
        mode="pink"
        themes="[object Object]"
      >
        Notifications
      </a>
    </li>
    <li>
      <button
        class="sc-bdfBwQ bftKdb"
        data-testid="signout"
        mode="pink"
        type="button"
      >
        Sign Out
      </button>
    </li>
  </nav>
</DocumentFragment>
`;
